### PR TITLE
Faster (re)runs & extended tests

### PR DESCRIPTION
**Pull request by siebrenf**
`Genomepy install` no longer overwriting files that have already been downloaded/generated, unless the --force flag is passed (implements #39)

(This does not automatically implement #38, but without overwriting, the genomepy command can now be run a second time, and will only get the remaining files.)

`Genomepy install` now tries to download primary assemblies before downloading the toplevel assemblies, unless the --toplevel flag is passed (implements #45)

Also incorporated the fix for #48, including the comments.

### CODE REVIEW

File: genomepy/functions.py [hunk 1]
```diff
@@ -154,7 +155,9 @@ def install_genome(
     genome_dir=None,
     localname=None,
     mask="soft",
+    toplevel=False,
```
Comment (simonvh) [inline genomepy/functions.py]: toplevel is probably specific top Ensembl (maybe also NCBI)? In that case we should not have it as an argument to `install_genome()` as it is provider-specific. Maybe use a `**kwargs` construct?

File: genomepy/functions.py [hunk 2]
```diff
@@ -290,21 +307,24 @@ def generate_env(fname=None):
                 fout.write("{}\n".format(env))
 
 
-def glob_fa_files(dirname):
+def glob_fa_files(dirname, ext='fa'):
```
Comment (simonvh) [inline genomepy/functions.py]: Update function name, as it is now not fa-specific. `glob_ext_files()`?

File: genomepy/functions.py (hunk 1 above)
Comment (simonvh) [reply genomepy/functions.py]: Thinking about this some more.. This warrants a general solution. What to do with provider-specific options. There is an additional option for Ensembl, `version`, which is currently not exposed in the command line tool. We need a general solution that ideally passes all provider-specific option through to the provider-specific functions.


For instance, at the command line:

```
genomepy install genome ensembl --ensembl-version 92 --ensembl-toplevel
```

Then all arguments that start with the provider name would be passed to that provider. The provider should provide these extra options, which could then be retrieved in  `genomepy/cli.py`.

File: genomepy/functions.py (hunk 2 above)
Comment (siebrenf) [reply genomepy/functions.py]: makes sense!

File: genomepy/functions.py (hunk 1 above)
Comment (siebrenf) [reply genomepy/functions.py]: The human genome on NCBI is stored in top-level format indeed ([source](ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13/README.txt)). Not sure if there is a primary assembly however. 

I will look into this next week.

Comment (siebrenf) [reply genomepy/functions.py]: I don't think this solution is supported _properly_ by `click`. Here's what I found:

Ideally we'd get a list of provider-specific options, and then iteratively add `@click.option()`s for each, thus providing a clear and consistent help menu. However, calling `@click` functions in a loop does not seem to work as it required all lines after `@click.command()` to start with either `@` or `def`.

One `@click.argument("provider-specific options", nargs=-1)` can accept any number of arguments, but the options wouldn't be documented. The explanation would have to be in the docstrings. The docstrings cannot have single newlines, so all options would either be in sequence or spaced with a whiteline between each. And the docstrings cannot be extended with functions.

The only remaining solution would be to add one `@click.option()` that accepts a string input. The help text of options can be extended using `.format()`, but now the options and their explanations would be squeezed on half of the screen.

Comment (simonvh) [reply genomepy/functions.py]: What about the options detailed here:

- https://stackoverflow.com/questions/50499340/specify-options-and-arguments-dynamically
- https://stackoverflow.com/questions/53147525/is-it-possible-to-dynamically-generate-commands-in-python-click

I think especially the first one would work. You can create a custom decorator for the `genomepy install` command.

Comment (siebrenf) [reply genomepy/functions.py]: Excellent suggestion. It worked! 

Current solution has a dictionary with all options in `cli.py`. This dictionary contains provider-specific options as well. 

About the idea to let each provider provide the options: I'm not sure how much cleaner this is. 

Right now, all options are in `cli.py`. When a new option is added, these can be added in `cli-py`, and extracted from `kwargs` in `providers.py`. No changes from the current workflow. 

However, if these were placed elsewhere (either in `providers.py` or even in a different location in `cli.py`), then the location for these options depends on whether they are general or specific. More messy in my opinion.

Comment (siebrenf) [reply genomepy/functions.py]: Opinion overruled :p 

Added provider specific options to `providers.py`!

**Comment by siebrenf** [comment]
left to do: 

- use tmp folders until downloads are complete

- remove gaps from the plugins and run it from utils by default.

- look into code coverage tests

**Comment by siebrenf** [comment]
Gaps has been taken out of the plugins. The code is now run after downloading the genome by default.

Inversely, sizes has been taken out of utils. The code was and still is only run in the plugin.

This "fixes" #49

File: README.md [hunk 3]
```diff
@@ -11,16 +11,15 @@
 Easily install and use genomes in Python and elsewhere!
 
 The goal is to have a _simple_ and _straightforward_ way to download and use genomic sequences. 
-Currently, genomepy supports UCSC, Ensembl and NCBI. 
+Currently, Genomepy supports UCSC, Ensembl and NCBI.
```
Comment (simonvh) [inline README.md]: I actually like the module lower-case. That means the package name and the "tool" name are identitical.

File: .flake8 [hunk 4]
```diff
@@ -1,6 +1,3 @@
-[metadata]
-description-file = README.md
```
Comment (simonvh) [inline .flake8]: Why is this removed?

File: README.md [hunk 5]
```diff
@@ -11,16 +11,15 @@
...
 
 [![asciicast](https://asciinema.org/a/eZttBuf5ly0AnjFVBiEIybbjS.png)](https://asciinema.org/a/eZttBuf5ly0AnjFVBiEIybbjS)
 
-**Pssst, hey there!** Is genomepy not doing what you want? Does it fail? Is it clunky? Is the documentation unclear? Have any other ideas on how to improve it? Don't be shy and [let us know](https://github.com/simonvh/genomepy/issues)!
+**Pssst, hey there!** Is Genomepy not doing what you want? Does it fail? Is it clunky? Is the documentation unclear? Have any other ideas on how to improve it? Don't be shy and [let us know](https://github.com/simonvh/genomepy/issues)!
 
 
 ## Installation
 
-Genomepy works with Python 3.4+.
```
Comment (simonvh) [inline README.md]: Leave in there, but maybe change to Python 3.6+?

File: setup.py [hunk 6]
```diff
@@ -21,41 +17,39 @@
...
-    "Programming Language :: Python :: 3.6",
+    "Programming Language :: Python :: 3",
     "Topic :: Scientific/Engineering :: Bio-Informatics",
 ]
 
 setup(
     name="genomepy",
     version=__version__,  # noqa: F821
-    description="Genomes in Python",
+    description="Download genomes in Python",
```
Comment (simonvh) [inline setup.py]: Maybe we need to change this. You can download genomes, but that's not specific to Python, and you can also *use* these genomes easily in Python. Come up with not too long sentence that covers this?

File: README.md [hunk 7]
```diff
@@ -33,10 +32,14 @@ Or via pip:
 $ pip install genomepy
 ```
 
-If you install via pip, you will have to install some dependencies, only if 
-you want to use the annotation download feature. 
-You will have to install the following
-utilities and make sure they are in your PATH:
+Installation via pip requires `pypandoc` installed. Additionally,
```
Comment (simonvh) [inline README.md]: Yeah, this kinda sucks.The problem is that pypandoc is necessary for only one thing: making sure the description on pypi is correct. Pypi used to only support rst README format. However, it should now support markdown (md) as well: https://pypi.org/project/markdown-description-example/ 
When I tried this previously it didn't work reliably, but we can try it again. (Worst case scenario: the README doesn't look nice on pypi).

File: setup.py (hunk 6 above)
Comment (siebrenf) [reply setup.py]: how about "Downloading & processing of genomes in command line and Python"?

File: README.md (hunk 7 above)
Comment (siebrenf) [reply README.md]: Since I cannot upload this repo to testPypi, I cannot check it. I'll leave this for you :)

If you decide to remove pypandoc, references to it are in the `README`, `setup.py` and the `enviroment.yml`.

Since I removed setup.cfg, the README might already be gone. Needs to be tested on testPypi too.
